src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,487 @@
+                    return CHIP_NO_ERROR;
+                }
+            }
+        }
+#endif // EMBER_AF_PLUGIN_ON_OFF
+
+        // todo should we move this before the OnMode section and check for boot reason in the onoff server code?
+        // If we have powered up because of an OTA, do not override the CurrentMode.
+        BootReasonType bootReason = BootReasonType::kUnspecified;
+        CHIP_ERROR error          = DeviceLayer::GetDiagnosticDataProvider().GetBootReason(bootReason);
+
+        if (error != CHIP_NO_ERROR)
+        {
+            ChipLogError(
+                Zcl, "Unable to retrieve boot reason: %" CHIP_ERROR_FORMAT ". Assuming that we did not reboot because of an OTA",
+                error.Format());
+            bootReason = BootReasonType::kUnspecified;
+        }
+
+        if (bootReason == BootReasonType::kSoftwareUpdateCompleted)
+        {
+            ChipLogDetail(Zcl, "ModeBase: StartUpMode is ignored for OTA reboot.");
+            return CHIP_NO_ERROR;
+        }
+
+        // Set CurrentMode to StartUpMode
+        if (mStartUpMode.Value() != mCurrentMode)
+        {
+            ChipLogProgress(Zcl, "ModeBase: Changing CurrentMode to the StartUpMode value.");
+            Status status = UpdateCurrentMode(mStartUpMode.Value());
+            if (status != Status::Success)
+            {
+                ChipLogError(Zcl, "ModeBase: Failed to change the CurrentMode to the StartUpMode value: %u", to_underlying(status));
+                return StatusIB(status).ToChipError();
+            }
+
+            ChipLogProgress(Zcl, "ModeBase: Successfully initialized CurrentMode to the StartUpMode value %u",
+                            mStartUpMode.Value());
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+template <typename RequestT, typename FuncT>
+void Instance::HandleCommand(HandlerContext & handlerContext, FuncT func)
+{
+    if (!handlerContext.mCommandHandled && (handlerContext.mRequestPath.mCommandId == RequestT::GetCommandId()))
+    {
+        RequestT requestPayload;
+
+        // If the command matches what the caller is looking for, let's mark this as being handled
+        // even if errors happen after this. This ensures that we don't execute any fall-back strategies
+        // to handle this command since at this point, the caller is taking responsibility for handling
+        // the command in its entirety, warts and all.
+        //
+        handlerContext.SetCommandHandled();
+
+        if (DataModel::Decode(handlerContext.mPayload, requestPayload) != CHIP_NO_ERROR)
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath,
+                                                     Protocols::InteractionModel::Status::InvalidCommand);
+            return;
+        }
+
+        func(handlerContext, requestPayload);
+    }
+}
+
+void Instance::handleChangeToMode(HandlerContext & ctx, const Commands::ChangeToMode::DecodableType & commandData)
```

Comment:
With the current pattern, this cannot be done because there are two versions of this function, one private and one not. I will be changing the pattern to reintroduce a delegate class after I go through all the comments. This change will fix it.